pyteal/ast/subroutine.py

```diff
@@ -248,9 +302,55 @@ def __call__(self, fnImplementation: Callable[..., Expr]) -> SubroutineFnWrapper
 
 
 def evaluateSubroutine(subroutine: SubroutineDefinition) -> SubroutineDeclaration:
-    argumentVars = [ScratchVar() for _ in range(subroutine.argumentCount())]
-    loadedArgs = [var.load() for var in argumentVars]
+    """
+    Puts together the data necessary to define the code for a subroutine.
+    "evaluate" is used here to connote evaluating the PyTEAL AST into a SubroutineDeclaration,
+    but not actually placing it at call locations. The trickiest part here is managing the subroutine's arguments.
+    The arguments are needed for two different code-paths, and there are 2 different argument types to consider
+    for each of the code-paths:
+
+    2 Argument Usages / Code-Paths
+    - -------- ------   ----------
+    Usage (A) for run-time: "argumentVars" --reverse--> "bodyOps"
+        These are "store" expressions that pick up parameters that have been pre-placed on the stack prior to subroutine invocation.
+        The argumentVars are stored into local scratch space to be used by the TEAL subroutine.
+
+    Usage (B) for compile-time: "loadedArgs"
+        These are expressions supplied to the user-defined PyTEAL function.
+        The loadedArgs are invoked to by the subroutine to create a self-contained AST which will translate into a TEAL subroutine.
+
+    In both usage cases, we need to handle
+
+    2 Argument Types
+    - -------- -----
+    Type 1 (by-value): these have python type Expr
+    Type 2 (by-reference): these hae python type ScratchVar
+
+    Usage (A) "argumentVars" - Storing pre-placed stack variables into local scratch space:
+        Type 1. (by-value) use ScratchVar.store() to pick the actual value into a local scratch space
+        Type 2. (by-reference) ALSO use ScratchVar.store() to pick up from the stack
+            NOTE: SubroutineCall.__teal__() has placed the _SLOT INDEX_ on the stack so this is stored into the local scratch space
+
+    Usage (B) "loadedArgs" - Passing through to an invoked PyTEAL subroutine AST:
+        Type 1. (by-value) use ScratchVar.load() to have an Expr that can be compiled in python by the PyTEAL subroutine
+        Type 2. (by-reference) use a DynamicScratchVar as the user will have written the PyTEAL in a way that satifies
+            the ScratchVar API. I.e., the user will write `x.load()` and `x.store(val)` as opposed to just `x`.
+    """
+
+    def var_n_loaded(param):
+        if param in subroutine.by_ref_args:
+            argVar = DynamicScratchVar(TealType.anytype)
+            loaded = argVar
+        else:
+            argVar = ScratchVar(TealType.anytype)
+            loaded = argVar.load()
+
+        return argVar, loaded
+
+    args = subroutine.arguments()
```

Comment:
I believe that one gets back on OrderedDict so that we should be safe here, but the concerns you raise are worthy of further investigation on my part.

Reply:
Results of further investigation:
We can see that in practice (running in Python 3.9) that this code converts `odict_keys` to `list`, so we're safe here:
```python
>>> from inspect import signature
>>> def foo(x, y: int, z: str):
...     pass
... 
>>> sig = signature(foo)
>>> params = sig.parameters
>>> params
mappingproxy(OrderedDict([('x', <Parameter "x">), ('y', <Parameter "y: int">), ('z', <Parameter "z: str">)]))
>>> params.keys()
odict_keys(['x', 'y', 'z'])
```

According to [python's official docs](https://docs.python.org/3/library/inspect.html#inspect.Signature.parameters):
> Parameters appear in strict definition order, including keyword-only parameters.

But they add a troubling caveat, that looks more fundamental _regardless_ of approach taken here:
> Changed in version 3.7: Python only explicitly guaranteed that it preserved the declaration order of keyword-only parameters as of version 3.7, although in practice this order had always been preserved in Python 3.